File: src/main/java/com/kingdom/util/computercardaction/ProsperityComputerCardActionHandler.java
package com.kingdom.util.computercardaction;

import com.kingdom.model.Card;
import com.kingdom.model.CardAction;
import com.kingdom.model.Game;
import com.kingdom.model.Player;
import com.kingdom.model.computer.ComputerPlayer;
import com.kingdom.util.cardaction.CardActionHandler;

import java.util.ArrayList;
import java.util.List;

public class ProsperityComputerCardActionHandler {
    public static void handleCardAction(CardAction cardAction, ComputerPlayer computer) {

        Game game = computer.getGame();
        Player player = computer.getPlayer();

        String cardName = cardAction.getCardName();
        int type = cardAction.getType();

        if (cardName.equals("Bishop")) {
            List<Integer> cardIds;
            if (computer.getNumCardsWorthTrashing(cardAction.getCards()) > 0) {
                cardIds = computer.getCardsToTrash(cardAction.getCards(), 1);
            } else {
                cardIds = computer.getCardsToDiscard(cardAction.getCards(), 1, false);
            }
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
        } else if (cardName.equals("Bishop 2")) {
            List<Integer> cardIds;
            if (computer.getNumCardsWorthTrashing(cardAction.getCards()) > 0) {
                cardIds = computer.getCardsToTrash(cardAction.getCards(), 1);
            } else {
                cardIds = new ArrayList<Integer>();
            }
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
        } else if (cardName.equals("Contraband")) {
            //todo better logic for determining card
            List<Integer> cardIds = new ArrayList<Integer>();
            Card card = computer.getHighestCostCard(cardAction.getCards());
            cardIds.add(card.getCardId());
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
        } else if (cardName.equals("Counting House")) {
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, null, null, null, cardAction.getEndNumber());
        } else if (cardName.equals("Expand")) {
            if (type == CardAction.TYPE_TRASH_CARDS_FROM_HAND) {
                Card cardToTrash = computer.getLowestCostCard(cardAction.getCards());
                List<Integer> cardIds = new ArrayList<Integer>();
                if (cardToTrash != null) {
                    cardIds.add(cardToTrash.getCardId());
                }
                CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
            } else {
                Card cardToGain = computer.getHighestCostCard(cardAction.getCards());
                List<Integer> cardIds = new ArrayList<Integer>();
                if (cardToGain != null) {
                    cardIds.add(cardToGain.getCardId());
                }
                CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
            }
        } else if (cardName.equals("Forge")) {
            //todo need way better logic for this card
            if (type == CardAction.TYPE_TRASH_UP_TO_FROM_HAND) {
                int numToTrash = 3;
                if (cardAction.getNumCards() < 3) {
                    numToTrash = cardAction.getNumCards();
                }
                CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, computer.getCardsToTrash(cardAction.getCards(), numToTrash), null, null, -1);
            } else {
                Card cardToGain = computer.getHighestCostCard(cardAction.getCards());
                List<Integer> cardIds = new ArrayList<Integer>();
                if (cardToGain != null) {
                    cardIds.add(cardToGain.getCardId());
                }
                CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
            }
        } else if (cardName.equals("Goons")) {
            int numCardsToDiscard = player.getHand().size() - 3;
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, computer.getCardsToDiscard(cardAction.getCards(), numCardsToDiscard), null, null, -1);
        } else if (cardName.equals("King's Court")) {
            Card cardToPlay = computer.getActionToDuplicate(cardAction.getCards(), 3);
            if (cardToPlay == null) {
                cardToPlay = cardAction.getCards().get(0);
            }
            List<Integer> cardIds = new ArrayList<Integer>();
            cardIds.add(cardToPlay.getCardId());
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
        } else if (cardName.equals("Loan")) {
            String choice;
            if (cardAction.getCards().get(0).isCopper()) {
                choice = "trash";
            } else {
                choice = "discard";
            }
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, null, null, choice, -1);
        } else if (cardName.equals("Mint")) {
            Card cardToGain = computer.getHighestCostCard(cardAction.getCards());
            List<Integer> cardIds = new ArrayList<Integer>();
            if (cardToGain != null) {
                cardIds.add(cardToGain.getCardId());
            }
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
        } else if (cardName.equals("Mountebank")) {
            //todo determine when it would be good to get curse and copper
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, null, null, "discard", -1);
        } else if (cardName.equals("Rabble")) {
            //todo determine when to reorder
            List<Integer> cardIds = new ArrayList<Integer>();
            for (Card card : cardAction.getCards()) {
                cardIds.add(card.getCardId());
            }
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
        } else if (cardName.equals("Royal Seal")) {
            String yesNoAnswer;
            if (computer.isCardToDiscard(cardAction.getCards().get(0))) {
                yesNoAnswer = "no";
            } else {
                yesNoAnswer = "yes";
            }
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, null, yesNoAnswer, null, -1);
        } else if (cardName.equals("Trade Route")) {
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, computer.getCardsToTrash(cardAction.getCards(), cardAction.getNumCards()), null, null, -1);
        } else if (cardName.equals("Vault")) {
            if (type == CardAction.TYPE_DISCARD_UP_TO_FROM_HAND) {
                List<Integer> cardsToDiscard = new ArrayList<Integer>();
                for (Card card : cardAction.getCards()) {
                    if (computer.isCardToDiscard(card)) {
                        cardsToDiscard.add(card.getCardId());
                    }
                }
                CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardsToDiscard, null, null, -1);
            } else {
                String yesNoAnswer;
                int cardsToDiscard = 0;
                for (Card card : cardAction.getCards()) {
                    if (computer.isCardToDiscard(card)) {
                        cardsToDiscard++;
                    }
                }
                if (cardsToDiscard >= cardAction.getNumCards()) {
                    yesNoAnswer = "yes";
                } else {
                    yesNoAnswer = "no";
                }
                CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, null, yesNoAnswer, null, -1);
            }
        } else if (cardName.equals("Vault2")) {
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, computer.getCardsToDiscard(cardAction.getCards(), 2), null, null, -1);
        } else if (cardName.equals("Watchtower")) {
            String choice;
            if (computer.isCardToTrash(cardAction.getCards().get(0))) {
                choice = "trash";
            } else if (computer.isCardToDiscard(cardAction.getCards().get(0))) {
                choice = "no_reveal";
            } else {
                choice = "deck";
            }
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, null, null, choice, -1);
        } else {
            throw new RuntimeException("Prosperity Card Action not handled for card: " + cardAction.getCardName() + " and type: " + cardAction.getType());
        }
    }
}

File: src/main/java/com/kingdom/util/computercardaction/ProsperityComputerCardActionHandler.kt
package com.kingdom.util.computercardaction

import com.kingdom.model.CardAction
import com.kingdom.model.computer.ComputerPlayer
import com.kingdom.util.cardaction.CardActionHandler
import java.util.*

object ProsperityComputerCardActionHandler {
    fun handleCardAction(cardAction: CardAction, computer: ComputerPlayer) {

        val game = computer.game
        val player = computer.player

        val cardName = cardAction.cardName
        val type = cardAction.type

        when (cardName) {
            "Bishop" -> {
                val cardIds = if (computer.getNumCardsWorthTrashing(cardAction.cards) > 0) {
                    computer.getCardsToTrash(cardAction.cards, 1)
                } else {
                    computer.getCardsToDiscard(cardAction.cards, 1, false)
                }
                CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1)
            }
            "Bishop 2" -> {
                val cardIds = if (computer.getNumCardsWorthTrashing(cardAction.cards) > 0) {
                    computer.getCardsToTrash(cardAction.cards, 1)
                } else {
                    ArrayList()
                }
                CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1)
            }
            "Contraband" -> {
                //todo better logic for determining card
                val cardIds = ArrayList<Int>()
                val card = computer.getHighestCostCard(cardAction.cards)
                cardIds.add(card!!.cardId)
                CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1)
            }
            "Counting House" -> CardActionHandler.handleSubmittedCardAction(game, player, null!!, null, null, cardAction.endNumber)
            "Expand" -> when (type) {
                CardAction.TYPE_TRASH_CARDS_FROM_HAND -> {
                    val cardToTrash = computer.getLowestCostCard(cardAction.cards)
                    val cardIds = ArrayList<Int>()
                    if (cardToTrash != null) {
                        cardIds.add(cardToTrash.cardId)
                    }
                    CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1)
                }
                else -> {
                    val cardToGain = computer.getHighestCostCard(cardAction.cards)
                    val cardIds = ArrayList<Int>()
                    if (cardToGain != null) {
                        cardIds.add(cardToGain.cardId)
                    }
                    CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1)
                }
            }
            "Forge" -> //todo need way better logic for this card
                when (type) {
                    CardAction.TYPE_TRASH_UP_TO_FROM_HAND -> {
                        var numToTrash = 3
                        if (cardAction.numCards < 3) {
                            numToTrash = cardAction.numCards
                        }
                        CardActionHandler.handleSubmittedCardAction(game, player, computer.getCardsToTrash(cardAction.cards, numToTrash), null, null, -1)
                    }
                    else -> {
                        val cardToGain = computer.getHighestCostCard(cardAction.cards)
                        val cardIds = ArrayList<Int>()
                        if (cardToGain != null) {
                            cardIds.add(cardToGain.cardId)
                        }
                        CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1)
                    }
                }
            "Goons" -> {
                val numCardsToDiscard = player.hand.size - 3
                CardActionHandler.handleSubmittedCardAction(game, player, computer.getCardsToDiscard(cardAction.cards, numCardsToDiscard), null, null, -1)
            }
            "King's Court" -> {
                var cardToPlay = computer.getActionToDuplicate(cardAction.cards, 3)
                if (cardToPlay == null) {
                    cardToPlay = cardAction.cards[0]
                }
                val cardIds = ArrayList<Int>()
                cardIds.add(cardToPlay.cardId)
                CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1)
            }
            "Loan" -> {
                val choice = if (cardAction.cards[0].isCopper) {
                    "trash"
                } else {
                    "discard"
                }
                CardActionHandler.handleSubmittedCardAction(game, player, null!!, null, choice, -1)
            }
            "Mint" -> {
                val cardToGain = computer.getHighestCostCard(cardAction.cards)
                val cardIds = ArrayList<Int>()
                if (cardToGain != null) {
                    cardIds.add(cardToGain.cardId)
                }
                CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1)
            }
            "Mountebank" -> //todo determine when it would be good to get curse and copper
                CardActionHandler.handleSubmittedCardAction(game, player, null!!, null, "discard", -1)
            "Rabble" -> {
                //todo determine when to reorder
                val cardIds = cardAction.cards.map { it.cardId }
                CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1)
            }
            "Royal Seal" -> {
                val yesNoAnswer = if (computer.isCardToDiscard(cardAction.cards[0])) {
                    "no"
                } else {
                    "yes"
                }
                CardActionHandler.handleSubmittedCardAction(game, player, null!!, yesNoAnswer, null, -1)
            }
            "Trade Route" -> CardActionHandler.handleSubmittedCardAction(game, player, computer.getCardsToTrash(cardAction.cards, cardAction.numCards), null, null, -1)
            "Vault" -> when (type) {
                CardAction.TYPE_DISCARD_UP_TO_FROM_HAND -> {
                    val cardsToDiscard = cardAction.cards
                            .filter { computer.isCardToDiscard(it) }
                            .map { it.cardId }
                    CardActionHandler.handleSubmittedCardAction(game, player, cardsToDiscard, null, null, -1)
                }
                else -> {
                    val cardsToDiscard = cardAction.cards.count { computer.isCardToDiscard(it) }
                    val yesNoAnswer = if (cardsToDiscard >= cardAction.numCards) {
                        "yes"
                    } else {
                        "no"
                    }
                    CardActionHandler.handleSubmittedCardAction(game, player, null!!, yesNoAnswer, null, -1)
                }
            }
            "Vault2" -> CardActionHandler.handleSubmittedCardAction(game, player, computer.getCardsToDiscard(cardAction.cards, 2), null, null, -1)
            "Watchtower" -> {
                val choice = if (computer.isCardToTrash(cardAction.cards[0])) {
                    "trash"
                } else if (computer.isCardToDiscard(cardAction.cards[0])) {
                    "no_reveal"
                } else {
                    "deck"
                }
                CardActionHandler.handleSubmittedCardAction(game, player, null!!, null, choice, -1)
            }
            else -> throw RuntimeException("Prosperity Card Action not handled for card: " + cardAction.cardName + " and type: " + cardAction.type)
        }
    }
}
